### ISSUE TITLE
CHARGE_ACCOUNT no longer allows for sub-projects

### ISSUE THREAD

**Issue by jhkennedy** [ESMCI/cime#2886]
When using the `-p/--project` option in CIME, it is assumed that the account to charge for compute time and project directory are the same. However, on machines like Titan and Summit, time often needs to be charged to a subproject like `cli106ice`, but the project directory doesn't reflect the subproject name (`/lustre/atlas/proj-shared/cli106ice/` is not a valid path, and instead should be `/lustre/atlas/proj-shared/cli106/`). This is discussed in issue E3SM-Project/E3SM#1588

PR E3SM-Project/E3SM#1718 added an optional `CHARGE_ACCOUNT` which allow for charging sub-projects while using the `PROJECT` name for directory paths. 

However, issue ESMCI/cime#2676 thought that `CHARGE_ACCOUNT` and `PROJECT` should always be the same, and PR ESMCI/cime#2763 removed the ability to use subprojects. 

Because Summit will be a supported machine, and Titan is a machine currently used,  ESMCI/cime#2763 should be reverted. 

@billsacks , @sarich

**Comment by billsacks** [comment]
@jhkennedy - I may be remembering/understanding wrong, but I don't think that simply reverting #2763 is the right answer here. According to the `create_newcase` documentation:

```
  --project PROJECT, -project PROJECT
                        Specify a project id as used in batch system accounting.
```

yet without #2763 setting `--project` has no effect on the account code used, at least on some machines. Can you state your suggested user interface for setting both project and charge accounts? I'd argue (fairly strongly) that any change should preserve the ability for a simple setting of the `--project` argument to set all necessary xml variables correctly on machines for which there is no distinction between project and charge_account (in order to maintain backwards compatibility with users' scripts that just set `--project`).

**Comment by jhkennedy** [comment]
@billsacks you're definitely right about the lack of documentation of how these two options interplay -- I'll update that for sure for whatever fix we come up with.

**Comment by billsacks** [comment]
@jhkennedy I'm pretty sure that, in my case, I had project set in some way (either in my environment or via a .cesm_proj file), but _not_ CHARGE_ACCOUNT. Prior to #2763 , the logic for CHARGE_ACCOUNT didn't work exactly as it was documented: The actual operation of step (4) in the hierarchy should have been stated (more precisely) as: "default to value given by get_project". Note that `get_project` does not itself know anything about the `--project` command-line flag.

So maybe a way to fix this is to back out #2763 but to change the logic at the end of `get_charge_account` so that it considers a possible `--project` command-line argument. If I were doing this, I'd probably make the `--project` command-line argument be an argument to both `get_project` and `get_charge_account`, moving some of the logic from `case.py` into `get_project`. So `get_project` would first see if `project_arg` is set (not None); if so, set project based on that. `get_charge_account` would accept a `project_arg` and just pass it along to `get_project` (which would effectively set CHARGE_ACCOUNT to what's set using the `--project` flag if CHARGE_ACCOUNT isn't otherwise given). I haven't thought that through carefully, though, so I may be missing something.

**Comment by billsacks** [comment]
Because a code diff is worth 1000 words, here's what I meant, without the changes that would be needed on the caller end, and without any added documentation:

```diff
diff --git a/scripts/lib/CIME/utils.py b/scripts/lib/CIME/utils.py
index 2cf4464c6..f521de851 100644
--- a/scripts/lib/CIME/utils.py
+++ b/scripts/lib/CIME/utils.py
@@ -890,7 +890,7 @@ def get_timestamp(timestamp_format="%Y%m%d_%H%M%S", utc_time=False):
         time_tuple = time.localtime()
     return time.strftime(timestamp_format, time_tuple)

-def get_project(machobj=None):
+def get_project(machobj=None, project_flag=None):
     """
     Hierarchy for choosing PROJECT:
     0. Command line flag to create_newcase or create_test
@@ -900,6 +900,9 @@ def get_project(machobj=None):
     4  File $HOME/.cesm_proj         (this is for backward compatibility)
     5  config_machines.xml (if machobj provided)
     """
+    if project_flag is not None:
+        logger.info("Using project from command-line flag: " + project_flag)
+        return project_flag
     project = os.environ.get("PROJECT")
     if (project is not None):
         logger.info("Using project from env PROJECT: " + project)
@@ -936,7 +939,7 @@ def get_project(machobj=None):
     logger.info("No project info available")
     return None

-def get_charge_account(machobj=None):
+def get_charge_account(machobj=None, project_flag=None):
     """
     Hierarchy for choosing CHARGE_ACCOUNT:
     1. Environment variable CHARGE_ACCOUNT
@@ -975,7 +978,7 @@ def get_charge_account(machobj=None):
             return charge_account

     logger.info("No charge_account info available, using value from PROJECT")
-    return get_project(machobj)
+    return get_project(machobj=machobj, project_flag=project_flag)

 def find_files(rootdir, pattern):
     """
```

**Comment by jhkennedy** [comment]
@billsacks Ah, I see the problem! (Note: below is looking at the code after the revert)

So here: https://github.com/ESMCI/cime/blob/e19de7babd0d86824f009ef90839d36869f8d7bf/scripts/lib/CIME/case/case.py#L888-L891

`project` is the project that was set at the command line, and is then set with `self.set_value(...)`. Then the charge account is set like: https://github.com/ESMCI/cime/blob/e19de7babd0d86824f009ef90839d36869f8d7bf/scripts/lib/CIME/case/case.py#L895-L897

Which if we look at `get_charge_account()`, it *looks* like the charge account should be set to the project given at the command line here: https://github.com/ESMCI/cime/blob/e19de7babd0d86824f009ef90839d36869f8d7bf/scripts/lib/CIME/utils.py#L977-L978

BUT, the `self.set_value('PROJECT', project)` is not setting the value in `machobj` which `get_project()` looks in if it can't find it as an environment variable, in the cime config, or user config. So, as you said `get_charge_account` never "sees" the command line flag. You're fix would work, but I think this would be a little simpler: 

```diff
diff --git a/scripts/lib/CIME/case/case.py b/scripts/lib/CIME/case/case.py
index 29f9edb16..e98a71a17 100644
--- a/scripts/lib/CIME/case/case.py
+++ b/scripts/lib/CIME/case/case.py
@@ -892,7 +892,7 @@ class Case(object):
         elif machobj.get_value("PROJECT_REQUIRED"):
             expect(project is not None, "PROJECT_REQUIRED is true but no project found")
         # Get charge_account id if it exists
-        charge_account = get_charge_account(machobj)
+        charge_account = get_charge_account(machobj, project)
         if charge_account is not None:
             self.set_value("CHARGE_ACCOUNT", charge_account)
 
diff --git a/scripts/lib/CIME/utils.py b/scripts/lib/CIME/utils.py
index 2cf4464c6..a257b37a7 100644
--- a/scripts/lib/CIME/utils.py
+++ b/scripts/lib/CIME/utils.py
@@ -936,7 +936,7 @@ def get_project(machobj=None):
     logger.info("No project info available")
     return None
 
-def get_charge_account(machobj=None):
+def get_charge_account(machobj=None, project=None):
     """
     Hierarchy for choosing CHARGE_ACCOUNT:
     1. Environment variable CHARGE_ACCOUNT
@@ -975,7 +975,7 @@ def get_charge_account(machobj=None):
             return charge_account
 
     logger.info("No charge_account info available, using value from PROJECT")
-    return get_project(machobj)
+    return project
 
 def find_files(rootdir, pattern):
     """
```
Since `project` should already be `get_project(machobj)` if it wasn't given on the command line when `get_charge_account` is called, and `get_charge_account` isn't called anywhere else that I can see, there's no need to probe that again.

**Comment by jhkennedy** [comment]
@billsacks if my diff looks good to you, I'll fix the docs and open a PR.

**Comment by sarich** [comment]
@jhkennedy

#2763 forces the CHARGE_ACCOUNT to use the command-line --project value **only** if the separate charge account is never set. This is definitely not clear from the PR comments, but should now be correct in terms of the heirarchy.

Are you having an actual problem with the desired CHARGE_ACCOUNT getting overwritten, or just anticipating a problem from the comments?

In any case, I like your solution better.

**Comment by sarich** [comment]
No, you're right, #2763 doesn't do it right.

**Comment by jhkennedy** [comment]
@sarich yeah I wasn't able to get a job running on Titan today using a subproject which is how I found the issue. Tricky little problem to get both working!

**Comment by billsacks** [comment]
@jhkennedy yes, I agree that your solution is simpler, given the calling order. I'm happy with that - thanks.
